### PR TITLE
Sound null safety for framework and flutter_test

### LINKED ISSUE #68562: Additional type errors surfaced by enabling --sound-null-safety on framework unit tests
There are about 370 failures in the framework unit tests when enabling sound null safety, probably drastically fewer unique failures. The PR is https://github.com/flutter/flutter/pull/68536

The failures are: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8865951167182274320/+/steps/download_dependencies/0/stdout?format=raw

One fix is: https://github.com/flutter/engine/pull/21983

### PR DESCRIPTION
**Pull request by goderbauer**
This fixes all issues expect the test failures from `test/widgets/widget_inspector_test.dart`, which are resolved by @jacob314 in https://github.com/flutter/flutter/pull/68661.

Fixes https://github.com/flutter/flutter/issues/68562

### CODE REVIEW

File: examples/layers/analysis_options.yaml
Comment (jonahwilliams) [inline examples/layers/analysis_options.yaml:1]: Another option is to remove the layers example from the `flutter` framework tests and run them separtely

**Review by jonahwilliams — DISMISSED** [review]
Besides layer example LGTM

**Review by jonahwilliams — APPROVED** [review]
LGTM if LG to CI

**Comment by shihaohong** [comment]
CI is still sad: https://cirrus-ci.com/task/5469809754243072

**Comment by goderbauer** [comment]
The failing `widget_inspector_test.dart` are blocked on https://github.com/flutter/flutter/pull/68661.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by jonahwilliams** [comment]
The last few widget inspector failures need: 

```
diff --git a/packages/flutter_test/lib/src/_matchers_io.dart b/packages/flutter_test/lib/src/_matchers_io.dart
index 9c1f02244d..23c7a2b1a2 100644
--- a/packages/flutter_test/lib/src/_matchers_io.dart
+++ b/packages/flutter_test/lib/src/_matchers_io.dart
@@ -52,6 +52,8 @@ class MatchesGoldenFile extends AsyncMatcher {
     Future<ui.Image> imageFuture;
     if (item is Future<ui.Image>) {
       imageFuture = item;
+    } else if (item is Future<ui.Image?>) {
+      imageFuture = item.then((ui.Image? image) => image!);
     } else if (item is ui.Image) {
       imageFuture = Future<ui.Image>.value(item);
     } else {
```

Additionally you should only pass --sound-null-safety to the framework and flutter_test tests.

**Comment by jonahwilliams** [comment]
That works too :)

**Comment by shihaohong** [comment]
The failing internal tests are legit, but we can fix the code there after merging this PR -- Some user code expects the returned value to be a non-nullable type for the result of `Navigator.of(context)!.push()`.

File: dev/bots/test.dart
Comment (shihaohong) [inline dev/bots/test.dart:617]: I might've missed the context, but what about the layer example that's giving it  `mixedModeNullSafetyOptions` vs the rest of the tests?

File: examples/layers/rendering/custom_coordinate_systems.dart
Comment (shihaohong) [inline examples/layers/rendering/custom_coordinate_systems.dart:5]: Does this turn on null-safety, but in mixed mode for `examples/layers`?

File: packages/flutter/test/material/app_test.dart
Comment (shihaohong) [inline packages/flutter/test/material/app_test.dart:252]: The internal test failure is similar to what caused this test to fail

**Review by shihaohong — COMMENTED** [review]
The null-safety code changes themselves LGTM, just some questions about `example/layers` since I'm less familiar and might've missed some context.

File: dev/bots/test.dart
Comment (goderbauer) [reply dev/bots/test.dart:617]: Some files from this example are imported from the framework tests, so I've only migrated those files. The remaining files are still unmigrated, so the tests for the examples themselves need to run in mixed mode.

File: examples/layers/rendering/custom_coordinate_systems.dart
Comment (goderbauer) [reply examples/layers/rendering/custom_coordinate_systems.dart:5]: This allows us to use null safety syntax in this file (i.e. the `!` and `?`).

File: packages/flutter/test/material/app_test.dart
Comment (goderbauer) [reply packages/flutter/test/material/app_test.dart:252]: It's slightly different, fix is in cl/338282385.

**Comment by goderbauer** [comment]
Submitting, google3 is fixed with cl/338282385.

**Pull request merged**
